apiary-receiver/apiary-receiver-sqs/src/main/java/com/expedia/apiary/extensions/receiver/sqs/SqsMessageReader.java

```diff
@@ -0,0 +1,129 @@
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Iterator;
+import java.util.Optional;
+
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.model.DeleteMessageRequest;
+import com.amazonaws.services.sqs.model.Message;
+import com.amazonaws.services.sqs.model.ReceiveMessageRequest;
+
+import com.expedia.apiary.extensions.receiver.common.MessageReader;
+import com.expedia.apiary.extensions.receiver.common.error.SerDeException;
+import com.expedia.apiary.extensions.receiver.common.event.ListenerEvent;
+import com.expedia.apiary.extensions.receiver.sqs.messaging.SqsMessageDeserializer;
+
+public class SqsMessageReader implements MessageReader {
+  private static final Integer DEFAULT_POLLING_WAIT_TIME = 10;
+  private static final Integer DEFAULT_MAX_MESSAGES = 10;
+
+  private String queueUrl;
+  private int waitTimeSeconds;
+  private int maxMessages;
+  private SqsMessageDeserializer sqsMessageDeserializer;
+  private AmazonSQS consumer;
+  private Iterator<Message> records;
+
+  private SqsMessageReader(String queueUrl, int waitTimeSeconds, int maxMessages,
+                          SqsMessageDeserializer sqsMessageDeserializer, AmazonSQS consumer) {
+    this.queueUrl = queueUrl;
+    this.waitTimeSeconds = waitTimeSeconds;
+    this.maxMessages = maxMessages;
+    this.sqsMessageDeserializer = sqsMessageDeserializer;
+    this.consumer = consumer;
+  }
+
+  @Override
+  public void close() {
+    consumer.shutdown();
+  }
+
+  @Override
+  public Optional<ListenerEvent> read() {
+    if (records == null || !records.hasNext()) {
+      records = receiveMessage();
+    }
+    if (records.hasNext()) {
+      return Optional.of(eventPayLoad(records.next()));
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  public void delete(Message message) {
+    DeleteMessageRequest request = new DeleteMessageRequest()
+        .withQueueUrl(queueUrl)
+        .withReceiptHandle(message.getReceiptHandle());
+    consumer.deleteMessage(request);
+  }
+
+  private Iterator<Message> receiveMessage() {
+      ReceiveMessageRequest request = new ReceiveMessageRequest()
+          .withQueueUrl(queueUrl)
+          .withWaitTimeSeconds(waitTimeSeconds)
+          .withMaxNumberOfMessages(maxMessages);
+      return consumer.receiveMessage(request).getMessages().iterator();
+  }
+
+  private ListenerEvent eventPayLoad(Message message) {
+    try {
+      return sqsMessageDeserializer.unmarshal(message.getBody());
+    } catch (Exception e) {
+      throw new SerDeException("Unable to unmarshall event", e);
+    }
+  }
+
+  public static final class Builder {
+    private String queueUrl;
+    private Integer waitTimeSeconds;
+    private Integer maxMessages;
+    private AmazonSQS consumer;
+    private SqsMessageDeserializer sqsMessageDeserializer;
+
+    public Builder(String queueUrl, AmazonSQS consumer, SqsMessageDeserializer sqsMessageDeserializer) {
+      this.queueUrl = queueUrl;
```

Comment:
I think only the non-default arguments should go in the builder constructor. To me, that feels like only queue_url.

`AmazonSqs consumer = AmazonSQSClientBuilder.standard().withRegion(region(conf)).withCredentials(credentials(conf)).build()`

SqsMessageDeserializer sqsMessageDeserializer = new DefaultApiarySqsMessageDeserializer();

And then you will provide two methods in the builder to override these if required. a user should not need to worry about any of these things if he just want to use all defaults and read from a simple SQS queue.

Reply:
Sounds good. The default `consumer` will be slightly different though, as `conf` is no longer there: `new AmazonSQSClientBuilder.standard().build()`. This will use the `DefaultAwsRegionProviderChain` and `DefaultAWSCredentialsProviderChain` to find the region and credentials respectively.

Reply:
Yeah. That is even better then.

Reply:
Done